File: app/src/main/java/com/handstandsam/shoppingapp/di/RepositoryModule.java
package com.handstandsam.shoppingapp.di;

import android.content.Context;

import com.handstandsam.shoppingapp.network.ShoppingService;
import com.handstandsam.shoppingapp.repository.CategoryRepository;
import com.handstandsam.shoppingapp.repository.ItemRepository;
import com.handstandsam.shoppingapp.repository.UserRepository;

import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;

@Module
public class RepositoryModule {

    Context applicationContext;

    public RepositoryModule(Context applicationContext) {
        this.applicationContext = applicationContext;
    }

    @Singleton
    @Provides
    CategoryRepository categoryRepository() {
        return new CategoryRepository(applicationContext);
    }


    @Singleton
    @Provides
    ItemRepository itemRepository(ShoppingService shoppingService) {
        return new ItemRepository(shoppingService);
    }

    @Singleton
    @Provides
    UserRepository userRepository(ShoppingService shoppingService) {
        return new UserRepository(shoppingService);
    }
}

File: app/src/main/java/com/handstandsam/shoppingapp/di/RepositoryModule.kt
package com.handstandsam.shoppingapp.di

import android.content.Context
import com.handstandsam.shoppingapp.network.ShoppingService
import com.handstandsam.shoppingapp.repository.CategoryRepo
import com.handstandsam.shoppingapp.repository.ItemRepo
import com.handstandsam.shoppingapp.repository.UserRepo
import com.handstandsam.shoppingapp.repository.NetworkCategoryRepo
import com.handstandsam.shoppingapp.repository.NetworkItemRepo
import com.handstandsam.shoppingapp.repository.NetworkUserRepo
import dagger.Module
import dagger.Provides
import javax.inject.Singleton

@Module
class RepositoryModule(internal var applicationContext: Context) {

    @Singleton
    @Provides
    internal fun categoryRepository(shoppingService: ShoppingService): CategoryRepo {
        return NetworkCategoryRepo(shoppingService)
    }


    @Singleton
    @Provides
    internal fun itemRepository(shoppingService: ShoppingService): ItemRepo {
        return NetworkItemRepo(shoppingService)
    }

    @Singleton
    @Provides
    internal fun userRepository(shoppingService: ShoppingService): UserRepo {
        return NetworkUserRepo(shoppingService)
    }
}